elastic/beats
# Accept prefix as metric_types for stackdriver metricset in GCP

## What does this PR do?

This PR is to add accepting prefix using `metric_types` config parameter into `stackdriver` metricset for user to specify a prefix of the metric types instead of listing out all metric types one by one in the config.

## Why is it important?

This will make stackdriver configuration a lot shorter/simpler. For example, if you want to collect all instance metric types for compute service, you can use:
```
- module: googlecloud
  metricsets:
    - stackdriver
  period: 1m
  metrics:
    - aligner: ALIGN_NONE
      service: compute
      metric_types: 
        - "instance"
```
Instead of listing all metric types like below:
```
- module: googlecloud
  metricsets:
    - stackdriver
  period: 1m
  metrics:
    - aligner: ALIGN_NONE
      service: compute
      metric_types:
        - "instance/cpu/reserved_cores"
        - "instance/cpu/scheduler_wait_time"
        - "instance/cpu/usage_time"
        - "instance/cpu/utilization"
        - "instance/disk/max_read_bytes_count"
        - "instance/disk/max_read_ops_count"
        - "instance/disk/max_write_bytes_count"
        - "instance/disk/max_write_ops_count"
        - "instance/disk/read_bytes_count"
        - "instance/disk/read_ops_count"
        ......
```

## Checklist




Note: I need to spend some time on figuring out how to mock GCP APIs or add integration tests for GCP overall in a separate PR.

## How to test this PR locally
Enable `googlecloud` module and change googlecloud.yml to:
```
- module: googlecloud
  metricsets:
    - stackdriver
  zone: "europe-west1-c"
  project_id: elastic-observability
  credentials_file_path: "your JSON credentials file path"
  exclude_labels: false
  period: 1m
  metrics:
    - aligner: ALIGN_NONE
      service: compute
      metric_types: 
        - "instance/cpu"
```

Start Metricbeat and you should see these cpu metrics collected:
```
googlecloud.stackdriver.instance.cpu.reserved_cores.value
googlecloud.stackdriver.instance.cpu.usage_time.value
googlecloud.stackdriver.instance.cpu.utilization.value
```
Note: `googlecloud.stackdriver.instance.cpu.scheduler_wait_time` might not exist because it is only available for VMs that belong to the e2 family and it's for measuring the wait time for vCPU is ready to run but unexpectedly not scheduled to run.
